### PR TITLE
Adding proof of concept scale test purely in Robot Framework

### PR DESCRIPTION
**Pull request by kkhalasi**
This is a scale test find the max number of switches that can connect to one ONOS Controller. The test script does not create a cluster of nodes, but that can be easily implemented. 

Pre-conditions to running this test is Robot Framework + external robot libraries:
 - pip install pybot
 - pip install robotframework-ride
 - pip install robotframework-requests
 - pip install robotframework-sshlibrary
 - pip install robotframework-httplibrary
 - pip install robotframework-archivelibrary

Also, you must have a "test cell" as this test leverages the bash scripts that are used to install the controller. 

Usage eg. "pybot -v USER_HOME:/home/fedora -v ONOS_HOME:/opt/onos -v start:10 -v end:30 -v increments:10 ScalePOC.robot"

After the test complete, three files will be generate from where you issued the "pybot" command. The "report.html" file will be your results!

### CODE REVIEW

**Comment by jhall11** [comment]
Since this isn't using TestON, could you move this to:
ONLabTest/RobotTests/ScalePOC.robot

Can you also add a README file? This could be the what is in your first comment.

**Comment by kkhalasi** [comment]
Hi Jon,

I just moved the files.

Kailash

On Mon, Jun 29, 2015 at 5:35 PM, Jon Hall <notifications@github.com> wrote:

> Since this isn't using TestON, could you move this to:
> ONLabTest/RobotTests/ScalePOC.robot
>
> Can you also add a README file? This could be the what is in your first
> comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opennetworkinglab/ONLabTest/pull/138#issuecomment-116888207>
> .
>



-- 
*The information transmitted in this email, including any file attachments,
is intended only for the person(s) or entity to which it is addressed and
may contain confidential and/or privileged material. If you have received
this email in error please notify us immediately and delete the message
from your computer system. If you are not the intended recipient, you
are notified that printing, disclosing, copying, or distributing this
information is strictly prohibited. The sender cannot guarantee that this
email or any attachment to it is free of malicious code.  The sender
accepts no liability for any damage caused by any virus transmitted by this
email. *

**Comment by jhall11** [comment]
ok, Looks ok to me. We will try to run this test and see if I notice any issues.

**Comment by kkhalasi** [comment]
Let me know if you have any trouble running the test or if you need help
debugging. Any information will help me update the README file.

Thanks,
Kailash

On Tue, Jun 30, 2015 at 10:00 AM, Jon Hall <notifications@github.com> wrote:

> ok, Looks ok to me. We will try to run this test and see if I notice any
> issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opennetworkinglab/ONLabTest/pull/138#issuecomment-117256678>
> .
>



-- 
*The information transmitted in this email, including any file attachments,
is intended only for the person(s) or entity to which it is addressed and
may contain confidential and/or privileged material. If you have received
this email in error please notify us immediately and delete the message
from your computer system. If you are not the intended recipient, you
are notified that printing, disclosing, copying, or distributing this
information is strictly prohibited. The sender cannot guarantee that this
email or any attachment to it is free of malicious code.  The sender
accepts no liability for any damage caused by any virus transmitted by this
email. *

**Comment by kkhalasi** [comment]
Hi Jon,

Have you guys had a chance to run the Robot Test?

Thanks,
Kailash

On Tue, Jun 30, 2015 at 10:02 AM, Kailash Khalasi <kkhalasi@iix.net> wrote:

> Let me know if you have any trouble running the test or if you need help
> debugging. Any information will help me update the README file.
>
> Thanks,
> Kailash
>
> On Tue, Jun 30, 2015 at 10:00 AM, Jon Hall <notifications@github.com>
> wrote:
>
>> ok, Looks ok to me. We will try to run this test and see if I notice any
>> issues.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/opennetworkinglab/ONLabTest/pull/138#issuecomment-117256678>
>> .
>>
>
>
>
> --
> *The information transmitted in this email, including any file
> attachments, is intended only for the person(s) or entity to which it is
> addressed and may contain confidential and/or privileged material. If you
> have received this email in error please notify us immediately and delete
> the message from your computer system. If you are not the intended
> recipient, you are notified that printing, disclosing, copying,
> or distributing this information is strictly prohibited. The sender cannot
> guarantee that this email or any attachment to it is free of malicious
> code.  The sender accepts no liability for any damage caused by any
> virus transmitted by this email. *
>
>


-- 
*The information transmitted in this email, including any file attachments,
is intended only for the person(s) or entity to which it is addressed and
may contain confidential and/or privileged material. If you have received
this email in error please notify us immediately and delete the message
from your computer system. If you are not the intended recipient, you
are notified that printing, disclosing, copying, or distributing this
information is strictly prohibited. The sender cannot guarantee that this
email or any attachment to it is free of malicious code.  The sender
accepts no liability for any damage caused by any virus transmitted by this
email. *

File: TestON/RobotTests/ScalePOC.robot [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+Library           HttpLibrary.HTTP
+
+*** Variables ***
+##Grab the environment variables sources from your "cell"
+${CONTROLLER_IP}    %{OC1}
+${MININET_IP}    %{OCN}
+${CONTROLLER_USER}    %{ONOS_USER}
+${MININET_USER}    %{ONOS_USER}
+##USER_HOME used for public key
+${USER_HOME}    /home/fedora
```
Comment (jhall11) [inline TestON/RobotTests/ScalePOC.robot]: Maybe set this to %{HOME}?

File: TestON/RobotTests/README [hunk 2]
```diff
@@ -0,0 +1,16 @@
+This is a scale test find the max number of switches that can connect to one ONOS Controller. The test script does not create a cluster of nodes, but that can be easily implemented.
+
+Pre-conditions to running this test is Robot Framework + external robot libraries:
+
+pip install pybot
```
Comment (jhall11) [inline TestON/RobotTests/README]: I don't think this is needed. The only PyBot project I found with pip search was an irc bot and it was an externally hosted module. The pybot command seems to get installed when the robotframework modules are installed.

File: TestON/RobotTests/ScalePOC.robot [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+##ONOS_HOME is where the onos dist will be deployed on the controller vm
+${ONOS_HOME}    /opt/onos
+${RESTCONFPORT}    8181
+${LINUX_PROMPT}    $
+##SWITCHES_RESULT_FILE and JENKINS_WORKSPACE can be configurable...read overriding variables in README
+##SWITCHES_RESULT_FILE is used to plot data. you can use a jenkins post job for this or do it manually
+${SWITCHES_RESULT_FILE}    ${USER_HOME}/workspace/tools/switches.csv
+${JENKINS_WORKSPACE}    ${USER_HOME}/workspace/ONOS-Stable/
```
Comment (jhall11) [inline TestON/RobotTests/ScalePOC.robot]: The test initially failed until i created this folder. May want to note that somewhere

**Comment by kkhalasi** [comment]
Thanks for the comments Jon! I've updated the patch.

How were the results for you?

**Comment by jhall11** [comment]
It seemed a bit hard to debug when things went wrong. Some of the failure messages were almost useless For example, when I had an incorrect USER_HOME:

pybot -v start:10 -v end:30 -v increments:10 ScalePOC.robot
==============================================================================
ScalePOC :: ONOS Switch Scale Test
==============================================================================
Find Max Switches By Scaling :: Find the max number of switches fr... | FAIL |
OSError: [Errno 13] Permission denied: '/home/fedora'
------------------------------------------------------------------------------
ScalePOC :: ONOS Switch Scale Test                                    | FAIL |
Suite teardown failed:
Several failures occurred:

1) 1 != 0

2) 1 != 0

3) Given key file '/home/fedora/.ssh/id_rsa' does not exist.

4) AttributeError: 'NoneType' object has no attribute 'open_session'

5) AttributeError: 'NoneType' object has no attribute 'open_session'

6) Given key file '/home/fedora/.ssh/id_rsa' does not exist.

7) AttributeError: 'NoneType' object has no attribute 'open_session'

8) AttributeError: 'NoneType' object has no attribute 'open_session'

9) Given key file '/home/fedora/.ssh/id_rsa' does not exist.

10) AttributeError: 'NoneType' object has no attribute 'open_sftp_client'

1 critical test, 0 passed, 1 failed
1 test total, 0 passed, 1 failed
==============================================================================
Output:  /home/admin/output.xml
Log:     /home/admin/log.html
Report:  /home/admin/report.html




However, when using correct variables it seemed to run ok.

**Comment by kkhalasi** [comment]
I have added a comment before that variable and updated the README file to
show users how to override variables.

I'm trying to leverage the extremely useful ONOS bash scripts and those
require login with ssh keys.

Kailash

On Mon, Jul 6, 2015 at 1:53 PM, Jon Hall <notifications@github.com> wrote:

> It seemed a bit hard to debug when things went wrong. Some of the failure
> messages were almost useless For example, when I had an incorrect USER_HOME:
> pybot -v start:10 -v end:30 -v increments:10 ScalePOC.robot ScalePOC ::
> ONOS Switch Scale Test
>
> Find Max Switches By Scaling :: Find the max number of switches fr... |
> FAIL |
> OSError: [Errno 13] Permission denied: '/home/fedora'
>
> ScalePOC :: ONOS Switch Scale Test | FAIL |
> Suite teardown failed:
> Several failures occurred:
>
> 1) 1 != 0
>
> 2) 1 != 0
>
> 3) Given key file '/home/fedora/.ssh/id_rsa' does not exist.
>
> 4) AttributeError: 'NoneType' object has no attribute 'open_session'
>
> 5) AttributeError: 'NoneType' object has no attribute 'open_session'
>
> 6) Given key file '/home/fedora/.ssh/id_rsa' does not exist.
>
> 7) AttributeError: 'NoneType' object has no attribute 'open_session'
>
> 8) AttributeError: 'NoneType' object has no attribute 'open_session'
>
> 9) Given key file '/home/fedora/.ssh/id_rsa' does not exist.
>
> 10) AttributeError: 'NoneType' object has no attribute 'open_sftp_client'
>
> 1 critical test, 0 passed, 1 failed
> 1 test total, 0 passed, 1 failed
>
> Output: /home/admin/output.xml
> Log: /home/admin/log.html
> Report: /home/admin/report.html
>
> However, when using correct variables it seemed to run ok.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opennetworkinglab/ONLabTest/pull/138#issuecomment-118994890>
> .
>



-- 
*The information transmitted in this email, including any file attachments,
is intended only for the person(s) or entity to which it is addressed and
may contain confidential and/or privileged material. If you have received
this email in error please notify us immediately and delete the message
from your computer system. If you are not the intended recipient, you
are notified that printing, disclosing, copying, or distributing this
information is strictly prohibited. The sender cannot guarantee that this
email or any attachment to it is free of malicious code.  The sender
accepts no liability for any damage caused by any virus transmitted by this
email. *

**Pull request merged**